Backport: Elements - Merge element style and classname generation to single filter

aaronrobertshaw opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

Syncs the changes from https://github.com/WordPress/gutenberg/pull/59535 and https://github.com/WordPress/gutenberg/pull/59538.

These change update the elements block support filters such that the elements class name is only generated in a single location. This avoids conflicts between class names and makes the block support robust in the face of a block's data being filtered via other `render_block_data` filters.

To test:

Confirm https://github.com/WordPress/gutenberg/issues/59462 is still resolved

1. Edit a post and add a paragraph
2. Create a link with the paragraph
3. Via the block inspector apply a link color to the paragraph
4. Save and view the frontend confirming the correct link color is shown
5. Confirm nested applications of the same element styles work correctly:

Create some nested blocks containing links

1. Apply a set of element styles to the parent block
2. Apply different styles to a child block
3. Apply the first set of element styles to a grand-child block
4. Confirm the correct display of element styles

Unit tests:

```
npm run test:php -- --filter Tests_Block_Supports_WpRenderElementsSupport
npm run test:php -- --filter Tests_Block_Supports_WpRenderElementsSupportStyles
```

Trac ticket: https://core.trac.wordpress.org/ticket/60681

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

aaronrobertshaw committed:
Move element block support class and styler generation to render_block_data filter

aaronrobertshaw commented:
This PR is still only in a draft state as I have run out of time. Unfortunately, I'll be AFK until next week with limited bandwidth to push this across the line. 

Anyone is welcome to pick this up, push to it, or create an alternate PR 🙏

aaronrobertshaw committed:
Correct references to version in which deprecation occurs

aaronrobertshaw committed:
Fix copy pasta error

aaronrobertshaw committed:
Fix deprecation calls replacement

tellthemachines commented on src/wp-includes/block-supports/elements.php:
```diff
@@ -31,33 +33,43 @@ function wp_get_elements_class_name( $block ) {
 		),
 		'heading' => array(
-			'skip'  => wp_should_skip_block_supports_serialization( $block_type, 'color', 'heading' ),
+			'skip'  => isset( $options['link']['skip'] ) ? $options['heading']['skip'] : false,
```
Typo: `$options['link']['skip']` should be `$options['heading']['skip']`.

tellthemachines commented on src/wp-includes/block-supports/elements.php:
```diff
@@ -225,8 +246,37 @@ function wp_render_elements_support_styles( $pre_render, $block ) {
+ * Ensure the elements block support class name generated, and added to
+ * block attributes, in the `render_block_data` filter gets applied to the
+ * block's markup.
+ *
```
Minor: missing `@since 6.6.0`.

tellthemachines reviewed (commented):
This is testing well for me! In order to reproduce the bug that this fixes, I dumped the [filter that originally caused it in gutenberg](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php#L922) into the layout.php file. I was then able to confirm that this change fixes the issue.

Just a couple minor corrections below, but otherwise looks good!

aaronrobertshaw committed:
Fix typo

aaronrobertshaw committed:
Add missing @since 6.6.0

aaronrobertshaw commented:
Thanks for the review @tellthemachines 👍 

I've addressed the feedback https://github.com/WordPress/wordpress-develop/pull/6214/commits/1789ecf5fd659aa1da57c165cfc0086720c6e185 & https://github.com/WordPress/wordpress-develop/pull/6214/commits/8e60ca74fd4098e484488757b154e8e877610615.

tellthemachines reviewed (approved):
Thanks for the update! LGTM.

tellthemachines commented:
Committed in [r58074](https://core.trac.wordpress.org/changeset/58074).

mukeshpanchal27 commented on src/wp-includes/block-supports/elements.php:
```diff
@@ -101,50 +105,53 @@ function wp_render_elements_support( $block_content, $block ) {
+	 * `render_block_data` filter in 6.6.0 to avoid filtered attributes
+	 * breaking the application of the elements CSS class.
+	 *
+	 * @see https://github.com/WordPress/gutenberg/pull/59535.
```
Remove end period.

mukeshpanchal27 commented on src/wp-includes/block-supports/elements.php:
```diff
@@ -225,8 +246,38 @@ function wp_render_elements_support_styles( $pre_render, $block ) {
+ * @since 6.6.0
+ *
+ * @param  string $block_content Rendered block content.
+ * @param  array  $block         Block object.
+ *
+ * @return string                Filtered block content.
```
Remove empty line.

aaronrobertshaw commented:
@mukeshpanchal27 as these changes were already committed, I've created a follow-up PR to address the issues noted
- https://github.com/WordPress/wordpress-develop/pull/6488
